File: 백준/Platinum/1517. 버블 소트/버블 소트.py
n = int(input())
a = [*map(int, input().split())]

def sort(start: int, end: int) -> int:
    if start + 1 >= end:
        return 0
    
    mid = (start + end) // 2
    result = sort(start, mid) + sort(mid, end)

    global a
    arr_front = a[start:mid]
    len_front = mid - start
    ptr_front = 0
    arr_back = a[mid:end]
    len_back = end - mid
    ptr_back = 0
    i = start
    while ptr_front < len_front and ptr_back < len_back:
        if arr_front[ptr_front] > arr_back[ptr_back]:
            a[i] = arr_back[ptr_back]
            result += len_front - ptr_front
            ptr_back += 1
        else:
            a[i] = arr_front[ptr_front]
            ptr_front += 1
        i += 1
    
    while ptr_front < len_front:
        a[i] = arr_front[ptr_front]
        ptr_front += 1
        i += 1

    while ptr_back < len_back:
        a[i] = arr_back[ptr_back]
        ptr_back += 1
        i += 1

    return result

print(sort(0, n))

File: 백준/Platinum/1517. 버블 소트/버블 소트.c
#include <stdio.h>
#include <stdlib.h>
typedef long long int lld;

lld swap(int start, int end, lld *arr)
{
    if (start + 1 >= end)
    {
        return 0;
    }

    int mid = (start + end) / 2;
    lld result = swap(start, mid, arr) + swap(mid, end, arr);

    int len_front = mid - start;
    int len_back = end - mid;
    lld *arr_front = (lld*)malloc(sizeof(lld) * len_front);
    lld *arr_back = (lld*)malloc(sizeof(lld) * len_back);
    int i;
    for (i = start; i < mid; i++)
    {
        arr_front[i - start] = arr[i];
    }
    for (i = mid; i < end; i++)
    {
        arr_back[i - mid] = arr[i];
    }

    int ptr_front = 0;
    int ptr_back = 0;
    i = start;
    while (ptr_front < len_front && ptr_back < len_back)
    {
        if (arr_front[ptr_front] > arr_back[ptr_back])
        {
            arr[i++] = arr_back[ptr_back++];
            result += len_front - ptr_front;
        }
        else
        {
            arr[i++] = arr_front[ptr_front++];
        }
    }

    while (ptr_front < len_front)
    {
        arr[i++] = arr_front[ptr_front++];
    }
    while (ptr_back < len_back)
    {
        arr[i++] = arr_back[ptr_back++];
    }
    
    free(arr_front);
    free(arr_back);
    return result;
}

int main()
{
    int n;
    scanf("%d", &n);
    lld *a = (lld*)malloc(sizeof(lld) * n);
    for (int i = 0; i < n; i++)
    {
        scanf("%lld", a + i);
    }

    printf("%lld", swap(0, n, a));

    free(a);
    return 0;
}